contracts/state/Smt.sol

```diff
@@ -0,0 +1,454 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.15;
+pragma abicoder v2;
+
+import "../lib/Poseidon.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+uint256 constant MAX_SMT_DEPTH = 32;
+
+struct Proof {
+    uint256 root;
+    uint256[MAX_SMT_DEPTH] siblings;
+    uint256 oldKey;
+    uint256 oldValue;
+    bool isOld0;
+    uint256 key;
+    uint256 value;
+    uint256 fnc;
+}
+
+contract Smt is OwnableUpgradeable {
+    PoseidonUnit2 _poseidonUnit2;
+    PoseidonUnit3 _poseidonUnit3;
+
+    address internal _writer;
+
+    /**
+     * @dev Throws if called by any account other than the state contract.
+     */
+    modifier onlyWriter() {
+        require(_writer == _msgSender(), "caller has no permissions");
+        _;
+    }
+
+    enum NodeType {
+        EMPTY,
+        LEAF,
+        MIDDLE
+    }
+
+    /**
+     * @dev Struct saved information about SMT root change.
+     * @param RootHistory historical tree root.
+     * @param BlockTimestamp commit time when state was saved into blockchain.
+     * @param BlockN commit number of block when state was created.
+     */
+    struct RootHistoryInfo {
+        uint256 RootHistory;
+        uint64 BlockTimestamp;
+        uint64 BlockN;
+    }
+
+    struct Node {
```

Comment:
add comments for Node, NodeType

Reply:
Done. Added comments for all structs, enums and public methods as well.